infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java

```diff
@@ -153,17 +153,47 @@ public void reset(long topologyId, long snapshot) {
      */
     private void processUpdatesShadowStream(List<SMREntry> smrEntries, Long currentSeqNum, UUID shadowStreamUuid,
                                             UUID snapshotSyncId) {
-        CorfuStoreMetadata.Timestamp timestamp;
+        CorfuStoreMetadata.Timestamp timestamp = null;
+        List<SMREntry> buffer = new ArrayList<>();
+        long bufferSize = 0;
 
-        try (TxnContext txn = metadataManager.getTxnContext()) {
-            updateLog(txn, smrEntries, shadowStreamUuid);
-            ReplicationMetadata metadata = metadataManager.queryReplicationMetadata(txn, session);
-            ReplicationMetadata updatedMetadata = metadata.toBuilder().setLastSnapshotStarted(srcGlobalSnapshot)
+        for (SMREntry smrEntry : smrEntries) {
+            if (bufferSize + smrEntry.getSerializedSize() + metadataManager.getReplicationMetadata(session).getSerializedSize()
+                    > replicationContext.getConfig(session).getMaxApplySize()
+                    || maxEntriesLimitReached(shadowStreamUuid,
+                    buffer)) {
+                try (TxnContext txnContext = metadataManager.getTxnContext()) {
+                    updateLog(txnContext, buffer, shadowStreamUuid);
+                    ReplicationMetadata metadata = metadataManager.queryReplicationMetadata(txnContext, session);
+                    ReplicationMetadata updatedMetadata = metadata.toBuilder().setLastSnapshotStarted(srcGlobalSnapshot)
                     .setLastSnapshotTransferredSeqNumber(currentSeqNum)
                     .build();
 
-            metadataManager.updateReplicationMetadata(txn, session, updatedMetadata);
-            timestamp = txn.commit();
+                    metadataManager.updateReplicationMetadata(txnContext, session, updatedMetadata);
+                    txnContext.commit();
+                    buffer.clear();
+                    buffer.add(smrEntry);
+                    bufferSize = smrEntry.getSerializedSize() + updatedMetadata.getSerializedSize();
```

Comment:
The metadata's size is being added for every SMR entry. We don't need that since the metadata is updated only once per txn.

Here and in L180.
Instead, the metadata's size should be added only once... you can do it either at the beginning when the buffer is emtpy or you have the check as in the if condition on L161. 
Both is not needed.